### PR TITLE
Add STREAM for GPU and make necessary adjustments for it

### PR DESCRIPTION
**Pull request by e-kayrakli**
This PR adds a full Stream benchmark for GPU.

In order to achieve that it also makes the followign adjustments to the GPU
support:

- Adds proper grid size calculation based on loop "size" and block size.
  - This is currenlty done by passing an argument representing number of threads
    to the runtime's kernel launcher. From there, the runtime launcher does the
    computation.
  - So, this PR adjusts the runtime interface slightly, as well.
- Adds an early return check in the gpu kernel, in case the local thread index
  is out-of-bounds for the loop.
- Adds a `--gpu-block-size` compiler flag to control the block size of gpu
  kernels.
- Adjusts the denormalize pass to avoid replacing temps used for kernel launches
  with equivalent expressions. This is done to avoid making more significant
  adjustments in the kernel launch codegen.

While there:
- Moves the `Kernel launcher called` output earlier to catch fatal launch errors
  that are due to not being able to load a kernel from the fatbinary.
- Drop an unused variable from a test.

#### Test
- [x] test/gpu/native
- [x] standard

### CODE REVIEW

File: test/gpu/native/streamPrototype/stream.chpl [hunk 1]
```diff
@@ -0,0 +1,211 @@
...
+
+//
+// The program entry point
+//
+proc main() {
+  printConfiguration();   // print the problem size, number of trials, etc.
+
+  on here.getChild(1) {
+    //
+    // ProblemSpace describes the index set for the three vectors.  It
```
Comment (daviditen) [inline test/gpu/native/streamPrototype/stream.chpl:61]: This comment should be updated to not talk about Block distribution. Or just removed altogether.

File: test/gpu/native/streamPrototype/stream.chpl [hunk 2]
```diff
@@ -0,0 +1,211 @@
...
+    // is a 1D domain that is distributed according to a Block
+    // distribution.  In this case, the Block distribution is 1D
+    // distribution computed by blocking the bounding box 1..m across
+    // the set of locales.  The ProblemSpace domain also contains the
+    // indices 1..m.
+    //
+    const ProblemSpace = {1..m};
+
+    //
+    // A, B, and C are the three distributed vectors, declared to store
```
Comment (daviditen) [inline test/gpu/native/streamPrototype/stream.chpl:71]: Can you remove `distributed` from this comment?

**Review by daviditen — APPROVED** [review]
Looks good, thanks!

File: test/gpu/native/streamPrototype/stream.chpl (hunk 1 above)
Comment (e-kayrakli) [reply test/gpu/native/streamPrototype/stream.chpl:61]: I adjusted this comment and the one below.

**Comment by bradcray** [comment]
@e-kayrakli : What is it about `--gpu-block-size` that makes it a compile-time argument as opposed to an execution-time / config-style argument?  I.e., is the compiler actually doing anything with it apart from passing it along to some runtime routines or generating code against it symbolically?  Thanks!

**Comment by bradcray** [comment]
Potentially answering it myself:  This is the block size that indicates the "width" of the kernels rather than the block size of any data transfers, say, so maybe the compiler needs to know this value to create the right "width" of generated code(?).

**Comment by e-kayrakli** [comment]
First, yes, this is the width of the kernels. Second, the short answer is there is nothing deep here and it could have been (and still can be) an execution time thing.

More: there's nothing particular that necessitates blockSize to be a compile-time constant. However, where we want to be eventually is to have some sort of language feature letting the user choose the block size *per* forall/foreach. In that world, we will need the compiler to feed that size to the runtime. So, thinking that, I wrote the outliner hard-coded with 512 as the size and fed it to the runtime directly. Then, to explore its impact on STREAM performance I tied it to a compiler flag. If I was doing things today, maybe that value should have been a `config const` the compiler can pass to all kernel launch calls to the runtime?

(One of the things that I was thinking about or maybe confused by was that if you're writing CUDA code, this value must be a compile time constant. But, I think it is to be able to support the "nice" kernel launch syntax with `<<< >>>`. We are using the low-level CUDA Driver API, which takes kernel dimensions as just regular `unsigned int`s.)

Thinking about where we want to be, it would be a good idea to make it into a `config const` as the default block size for when the user omits it in code.

**Pull request merged**